lib/puppet/provider/yumrepo/inifile.rb

```diff
@@ -182,14 +182,27 @@ def self.repofiles
 
   # Build a virtual inifile by reading in numerous .repo files into a single
   # virtual file to ease manipulation.
+  # Will skip any inifile that is found to be currently managed as a file resource
+  # in the catalog.
   # @api private
   # @return [Puppet::Provider::Yumrepo::IniConfig::File] The virtual inifile representing
   #   multiple real files.
-  def self.virtual_inifile
+  def self.virtual_inifile(resources = nil)
+    current_yumrepo_inis = []
+    resources&.each do |key|
+      current_yumrepo_inis << key[1]
+    end
+
     unless @virtual
       @virtual = Puppet::Provider::Yumrepo::IniConfig::File.new
       repofiles.each do |file|
-        @virtual.read(file) if Puppet::FileSystem.file?(file)
+        index_virtual = true
+        unless current_yumrepo_inis.empty?
+          current_yumrepo_inis.each do |index|
+            index_virtual = false if index.catalog.resource('File', file)
+          end
```

Comment:
We can stop enumerating over `current_yumrepo_inis` as soon as we find the first `File` resource with path `file`. A common idiom to short-circuit is to do:

```diff
-          current_yumrepo_inis.each do |index|
-            index_virtual = false if index.catalog.resource('File', file)
-          end
+          index_virtual = !current_yumrepo_inis.any? { |index| index.catalog.resource('File', file) }